topics/releases.md

```diff
@@ -1,141 +1,108 @@
+We follow semantic versioning to provide explicit guarantees regarding backward compatibility.
 
-The first release candidate, or RC1, is released once it can be used for
-development purposes and for testing the new version. At this stage, most of
-the new features and changes the new version brings are ready for review, and
-the release's purpose is collecting the public's feedback.
+When discussing compatibility, we refer to the following API contracts:
 
-Subsequent release candidates are released every three weeks or so, primarily
-for fixing bugs. These may also add new features and introduce changes, but at
-a decreasing rate and decreasing potential risk towards the final release
-candidate.
+1. Valkey commands including their inputs, outputs, and defined behavior
+2. The functions and APIs that can be executed from a Lua script
+3. The RDB version
+4. The protocol used to establish and replicate data from primaries to replicas
+5. The protocol between nodes within a Valkey cluster
+6. The Valkey Module API interface
+7. The AOF on disk format
 
-### Stable tree
+### Patch versions
 
-Once development has ended and the frequency of critical bug reports for the
-release candidate wanes, it is ready for the final release. At this point, the
-release is marked as stable and is released with "0" as its patch-level
-version.
+*Patch* versions are released with backwards compatible bug fixes and should not introduce new features.
 
-## Versioning
+Upgrading from a previous patch version should be safe and seamless.
+It should be safe to run a primary-replica pair or a Valkey cluster with servers running on different patch versions.
 
-Stable releases liberally follow the usual `major.minor.patch` semantic
-versioning schema. The primary goal is to provide explicit guarantees regarding
-backward compatibility.
+*Patch* versions may also introduce small improvements such as performance or memory optimizations that are easy to verify as safe.
 
-### Patch-Level versions
+### Minor versions
 
-Patches primarily consist of bug fixes and very rarely introduce any
-compatibility issues.
+*Minor* versions are released with new functionality that is added in a backward compatible manner.
+Examples of new functionality include new commands, info fields, or configuration parameters.
 
-Upgrading from a previous patch-level version is almost always safe and
-seamless.
+Upgrading from a previous minor version should be safe, and will not introduce incompatibilities between servers in the cluster when default server configurations are used.
 
-New features and configuration directives may be added, or default values
-changed, as long as these don’t carry significant impacts or introduce
-operations-related issues.
+**NOTE:** Minor releases may include new commands and data types that can introduce incompatibility between servers in the cluster, but users need to opt-in to these features to cause this type of incompatibility.
```

Comment:
Here's an issue. Generally, when we introduce a new data type, the `rdb_version` is also updated accordingly. So, even if users haven't used the new data type, the newer version nodes in the cluster would already be utilizing the updated `rdb_version`. In this scenario, if data is migrated from a newer version to an older version, errors would occur.

Reply:
Yes. If we want to be forward-compatible, which I think is a good idea, we need to support writing RDB in an old version and a new version. We only need to be forward compatible one version, I think.

HLD: The first time a new datatype is used, we set a global flag. When we write the RDB, we check the flag. If the flag is set, we use the new RDB version. If the flag is not set, we use the older RDB version.

Reply:
@soloestoy the long term solution is AOF without RDB? valkey-io/valkey#59